cmd/internal/server/handlers/schema_aware_serializer.go

```diff
@@ -0,0 +1,228 @@
+		return l.Log(fivetransdk.LogLevel_SEVERE, fmt.Sprintf("marshal schema aware json serializer : %q", err))
+	}
+	return l.sender.Send(&fivetransdk.UpdateResponse{
+		Response: &fivetransdk.UpdateResponse_Operation{
+			Operation: &fivetransdk.Operation{
+				Op: &fivetransdk.Operation_Checkpoint{
+					Checkpoint: &fivetransdk.Checkpoint{
+						StateJson: string(state),
+					},
+				},
+			},
+		},
+	})
+}
+
+func (l *schemaAwareSerializer) Log(level fivetransdk.LogLevel, s string) error {
+	return l.sender.Send(&fivetransdk.UpdateResponse{
+		Response: &fivetransdk.UpdateResponse_LogEntry{
+			LogEntry: &fivetransdk.LogEntry{
+				Message: l.prefix + " " + s,
+				Level:   level,
+			},
+		},
+	})
+}
+
+func (l *schemaAwareSerializer) Record(result *sqltypes.Result, schema *fivetransdk.SchemaSelection, table *fivetransdk.TableSelection) error {
+	// make one response type per schema + table combination
+	// so we can avoid instantiating one object per table, and instead
+	// make one object per schema + table combo
+	if responseKey(schema, table) != l.recordResponseKey {
+		l.recordResponseKey = responseKey(schema, table)
+		l.recordResponse = &fivetransdk.UpdateResponse{
+			Response: &fivetransdk.UpdateResponse_Operation{
+				Operation: &fivetransdk.Operation{
+					Op: &fivetransdk.Operation_Record{
+						Record: &fivetransdk.Record{
+							SchemaName: &schema.SchemaName,
+							TableName:  table.TableName,
+							Type:       fivetransdk.OpType_UPSERT,
+							Data:       nil,
+						},
+					},
+				},
+			},
+		}
+
+		if _, ok := l.serializers[l.recordResponseKey]; !ok {
+
+			rs, err := generateRecordSerializer(table, schema.SchemaName, l.schemaList, l.serializeTinyIntAsBool)
+			if err != nil {
+				return err
+			}
+			l.serializers[l.recordResponseKey] = &rs
+
+		}
+	}
+
+	rs := *l.serializers[l.recordResponseKey]
+	rows, err := rs.Serialize(result)
+	if err != nil {
+		return l.Log(fivetransdk.LogLevel_SEVERE, fmt.Sprintf("record schema aware json serializer : %q", err))
+	}
+
+	for _, row := range rows {
+		operation, ok := l.recordResponse.Response.(*fivetransdk.UpdateResponse_Operation)
+		if !ok {
+			return l.Log(fivetransdk.LogLevel_SEVERE, fmt.Sprintf("recordResponse Operation is of type %T, not UpdateResponse_Operation", l.recordResponse.Response))
+		}
+		operationRecord, ok := operation.Operation.Op.(*fivetransdk.Operation_Record)
+		if !ok {
+			return l.Log(fivetransdk.LogLevel_SEVERE, fmt.Sprintf("recordResponse Operation.Op is of type %T not Operation_Record", operation.Operation.Op))
+		}
+		operationRecord.Record.Data = row
+		l.sender.Send(l.recordResponse)
+	}
+
+	return nil
+}
+
+func generateRecordSerializer(table *fivetransdk.TableSelection, selectedSchemaName string, schemaList *fivetransdk.SchemaList, serializeTinyIntAsBool bool) (recordSerializer, error) {
+	serializers := map[string]func(value sqltypes.Value) (*fivetransdk.ValueType, error){}
+	var err error
+	var tableSchema *fivetransdk.Table
```

Comment:
YOu probably want this inside the loop if you want to make sure you have a non-nil tableSchema value per schema in the schema list.